pool/src/pool/submit.rs

```diff
@@ -63,6 +61,87 @@ pub fn execute_submit(
     new_from_state.positions
 }
 
+// Note: if this looks like a good approach we should probably refactor the code
+// a bit to prevent duplicating execute_submit.
+pub fn execute_submit_with_flash_loan(
+    e: &Env,
+    from: &Address,
+    flash_loan: FlashLoan,
+    requests: Vec<Request>,
+    use_allowance: bool,
+) -> Positions {
+    if from == &e.current_contract_address()
+    {
+        panic_with_error!(e, &PoolError::BadRequest);
+    }
+    let mut pool = Pool::load(e);
+
+    // note: check_health is omitted since we always will want to check the health
+    // if a flash loan is involved.
+    let (actions, mut new_from_state, _) =
+        build_actions_from_request(e, &mut pool, from, requests);
+
+    // similarly to usual borrows, we want the flash loan to be debited before
+    // checking the health factor.
+    
+    // Warning: at this point user state is not written yet before invoking the 
+    // flash loan contract, so was this function ever reentrant this code would become
+    // unsafe.
+    {
+        let mut reserve = pool.load_reserve(e, &flash_loan.asset, true);
+        let d_tokens_minted = reserve.to_d_token_up(flash_loan.amount);
+        new_from_state.add_liabilities(e, &mut reserve, d_tokens_minted);
+        reserve.require_utilization_below_max(e);
+
+        e.events().publish(
+            (
+                Symbol::new(e, "flash_borrow"),
+                &flash_loan.asset,
+                &flash_loan.contract,
+            ),
+            (&flash_loan.amount, d_tokens_minted),
+        );
+    }
```

Comment:
The flash loan liabilities will need to be added before `build_actions_from_request`, right?

Reason being, it will be quite common for a `repay` request to be included of the flash loaned asset in the `requests` array.

`repay` has logic that reduces the repayment amount based on the actual liabilities on the block this is being processed on. So, the `flash_loan` liability needs to exist on the `Positions` object before `repay` is processed, that way the liability will accurately be reduced.